File: research/train_layer8.py

import torch
import torch.nn as nn
import torch.optim as optim
from torch.utils.data import Dataset, DataLoader
import numpy as np
import struct
import os

# Configuration
LAYER7_FILE = 'layer7.bin'
TARGET_FILE = 'akiyo_cif.yuv' # Use the CIF YUV as target
NUM_FRAMES = 50
CHANNELS = 56
ROWS = 144
COLS = 176
SCALE = 2
FILTERS_8 = 1
FILTER_SIZE_8 = 9
DROPOUT_RATE = 0.2 # Simulate race condition loss

class Layer8Dataset(Dataset):
    def __init__(self, layer7_path, target_path, num_frames, rows, cols, channels):
        self.rows = rows
        self.cols = cols
        self.channels = channels
        self.num_frames = num_frames
        
        # Load Layer 7 inputs (Features)
        # Format: double (8 bytes)
        # Size: num_frames * (rows * cols * channels)
        # We read all into memory (approx 50 * 144 * 176 * 56 * 8 bytes ≈ 500MB). OK.
        print("Loading input features...")
        with open(layer7_path, 'rb') as f:
            data = f.read()
            # shape (num_frames, channels, rows, cols) ? 
            # C code: imadd(img_fltr_p7 + i*rows*cols, ...)
            # i is channel index (0..55).
            # So memory layout is likely: Frame -> Channel -> Row -> Col ?
            # Wait, looking at source:
            # allocate `img_fltr_7` size `rows*cols*56`.
            # Access: `img_fltr_7 + j*rows*cols`. j is channel.
            # So consecutive memory is: Channel 0, Channel 1...
            # This matches PyTorch (C, H, W).
            self.inputs = np.frombuffer(data, dtype=np.float64).reshape(num_frames, channels, rows, cols)
            self.inputs = torch.from_numpy(self.inputs).float()
            
        # Load Target (HR Y-channel)
        print("Loading targets...")
        self.targets = []
        width_hr = cols * SCALE
        height_hr = rows * SCALE
        y_size = width_hr * height_hr
        uv_size = (width_hr // 2) * (height_hr // 2)
        frame_size = y_size + 2 * uv_size
        
        with open(target_path, 'rb') as f:
            for i in range(num_frames):
                # Seek to frame
                f.seek(i * frame_size)
                y_data = f.read(y_size)
                # Y data is uint8.
                y_img = np.frombuffer(y_data, dtype=np.uint8).reshape(height_hr, width_hr)
                # Normalize to 0-1? source.c outputs in 0-1 and then scales to 255.
                # source.c loads uint8 input, divides by 255.0.
                # So target should be 0-1?
                # Actually source.c Layer 8 output is added to `biases_layer8`.
                # And finally converted to uint8.
                # We should train against 0-255 or 0-1?
                # Inputs (Layer 7) are likely valid range.
                # Let's normalize target to 0-1 to match network scale likely.
                self.targets.append(y_img.astype(np.float32) / 255.0)
        
        self.targets = np.stack(self.targets)
        # Add channel dim: (N, 1, H, W)
        self.targets = torch.from_numpy(self.targets).unsqueeze(1)
        
        # Verify sizes
        print(f"Inputs: {self.inputs.shape}")
        print(f"Targets: {self.targets.shape}")

    def __len__(self):
        return self.num_frames

    def __getitem__(self, idx):
        return self.inputs[idx], self.targets[idx]

class RobustDeconvLayer(nn.Module):
    def __init__(self):
        super(RobustDeconvLayer, self).__init__()
        # We need to simulate summing 56 separate deconvs.
        # Standard ConvTranspose2d(56, 1, ...) sums internally.
        # To inject noise *during* summation, we act as if we have 56 separate filters.
        # We can use groups=1 (standard) and do trickery, OR use groups=56?
        # If groups=56, input 56 channels -> output 56 channels.
        # Each channel processed independently.
        # Then we Sum(outputs).
        # This allows us to apply Dropout to the 56 outputs before summing.
        
        self.deconv = nn.ConvTranspose2d(
            in_channels=56, 
            out_channels=56, 
            kernel_size=9, 
            stride=2, 
            padding=4, 
            output_padding=1,
            groups=56, # Independent channels
            bias=False # Bias is applied after sum
        )
        # Bias
        self.bias = nn.Parameter(torch.zeros(1))
        
        # Dropout
        self.dropout = nn.Dropout(p=DROPOUT_RATE)

    def forward(self, x):
        # x: (N, 56, 144, 176)
        # out: (N, 56, 288, 352)
        out = self.deconv(x)
        
        # Apply dropout to the channels (simulate lost race condition updates)
        # We want to drop entire channels? Or random pixels?
        # User said "race condition saat menjumlahkan 56 channel".
        # If using standard dropout, it drops elements independently.
        # This simulates fine-grained race condition (pixel level).
        out = self.dropout(out)
        
        # Sum across channels to get single Y channel
        # out: (N, 1, 288, 352)
        out = torch.sum(out, dim=1, keepdim=True)
        
        # Add bias
        out = out + self.bias
        return out

def save_weights(model, path):
    # Extract weights and save in format expected by source.c
    # source.c expects weights_layer8.txt
    # Loop i=0..num_filters8 (1)
    #   Loop j=0..num_channels8 (56)
    #     Loop kernel (9x9)
    # But we used groups=56.
    # self.deconv.weight shape: (56, 1, 9, 9) because (in_channels, out_channels/groups, k, k)
    # We essentially have 56 filters, each taking 1 input.
    # We need to map this to (1 filter, 56 channels).
    # It is effectively the same values.
    # channel j weight in PyTorch `deconv.weight[j, 0, :, :]` corresponds to `kernel` for channel j in source.c
    
    weights = model.deconv.weight.data.cpu().numpy() # (56, 1, 9, 9)
    bias = model.bias.data.cpu().numpy() # (1)
    
    print(f"Saving weights to {path}...")
    with open(path, 'w') as f:
        # source.c order:
        # Channel 0, Channel 1, ... Channel 55
        for j in range(56):
            k = weights[j, 0, :, :] # 9x9
            k_flat = k.flatten()
            for val in k_flat:
                f.write(f"{val:.10f}\n")
                
    print(f"Learned Bias: {bias[0]:.10f}")
    with open("bias_layer8_new.txt", "w") as f:
        f.write(f"{bias[0]:.10f}")

def main():
    if not os.path.exists(LAYER7_FILE):
        print(f"Error: {LAYER7_FILE} not found.")
        return
        
    dataset = Layer8Dataset(LAYER7_FILE, TARGET_FILE, NUM_FRAMES, ROWS, COLS, CHANNELS)
    dataloader = DataLoader(dataset, batch_size=4, shuffle=True)
    
    model = RobustDeconvLayer()
    
    # Initialize weights? Or load existing weights to start with?
    # Better to start from scratch or existing?
    # If we load existing, we fine tune.
    # Existing weights are in `weights/weights_layer8.txt`.
    # Let's try to load them if possible, but parsing them is work.
    # Random init is fine for this task usually.
    
    criterion = nn.MSELoss()
    optimizer = optim.Adam(model.parameters(), lr=1e-3)
    
    print("Starting training...")
    for epoch in range(10): # Quick training
        model.train()
        total_loss = 0
        for inputs, targets in dataloader:
            optimizer.zero_grad()
            outputs = model(inputs)
            loss = criterion(outputs, targets)
            loss.backward()
            optimizer.step()
            total_loss += loss.item()
        
        print(f"Epoch {epoch+1}, Loss: {total_loss/len(dataloader):.6f}")
        
    # Save
    save_weights(model, 'weights_layer8_robust.txt')
    print("Weights saved.")

if __name__ == "__main__":
    main()

File: research/train_layer8.c

#include <stdio.h>
#include <stdlib.h>
#include <math.h>
#include <time.h>

#define FRAMES 6
#define CHANNELS 56
#define ROWS 144
#define COLS 176
#define SCALE 2
#define KERNEL_SIZE 9
#define STRIDE 2
#define TARGET_ROWS (ROWS * SCALE)
#define TARGET_COLS (COLS * SCALE)
#define NUM_ITERATIONS 5000000
#define LR 0.0001
#define DROPOUT_RATE 0.3

// Global buffers
double *inputs; // FRAMES * CHANNELS * ROWS * COLS
double *targets; // FRAMES * TARGET_ROWS * TARGET_COLS
double weights[CHANNELS][KERNEL_SIZE][KERNEL_SIZE];
double bias = 0.0;

// Helper to get random double 0..1
double rand_double() {
    return (double)rand() / (double)RAND_MAX;
}

void load_data() {
    // Load Layer 7 inputs
    FILE *f = fopen("layer7.bin", "rb");
    if (!f) { printf("Error opening layer7.bin\n"); exit(1); }
    
    long input_size = (long)FRAMES * CHANNELS * ROWS * COLS;
    inputs = (double*)malloc(input_size * sizeof(double));
    if (!inputs) { printf("Memory alloc failed for inputs\n"); exit(1); }
    
    size_t read_count = fread(inputs, sizeof(double), input_size, f);
    if (read_count != input_size) { printf("Warning: Read %zu doubles needed %ld\n", read_count, input_size); }
    fclose(f);
    
    // Load Targets (Y channel of CIF YUV)
    f = fopen("akiyo_cif.yuv", "rb");
    if (!f) { printf("Error opening akiyo_cif.yuv\n"); exit(1); }
    
    long target_size = (long)FRAMES * TARGET_ROWS * TARGET_COLS;
    targets = (double*)malloc(target_size * sizeof(double));
    if (!targets) { printf("Memory alloc failed for targets\n"); exit(1); }
    
    int frame_size_y = TARGET_ROWS * TARGET_COLS;
    int frame_size_uv = (TARGET_ROWS/2) * (TARGET_COLS/2);
    int total_frame_bytes = frame_size_y + 2 * frame_size_uv;
    
    unsigned char *buf = (unsigned char*)malloc(frame_size_y);
    
    for (int i = 0; i < FRAMES; i++) {
        // Seek to Y start
        fseek(f, i * total_frame_bytes, SEEK_SET);
        fread(buf, 1, frame_size_y, f);
        
        for (int j = 0; j < frame_size_y; j++) {
            targets[i * frame_size_y + j] = (double)buf[j] / 255.0;
        }
    }
    free(buf);
    fclose(f);
    printf("Data loaded.\n");
}

void init_weights() {
    srand(time(NULL));
    for (int c = 0; c < CHANNELS; c++) {
        for (int i = 0; i < KERNEL_SIZE; i++) {
            for (int j = 0; j < KERNEL_SIZE; j++) {
                // Initialize small random weights
                weights[c][i][j] = (rand_double() - 0.5) * 0.1;
            }
        }
    }
    bias = 0.0;
}

void save_weights() {
    FILE *f = fopen("weights_layer8_robust.txt", "w");
    if (!f) return;
    
    // Format: Channel 0 (81 values), Channel 1...
    for (int c = 0; c < CHANNELS; c++) {
        for (int i = 0; i < KERNEL_SIZE; i++) {
            for (int j = 0; j < KERNEL_SIZE; j++) {
                fprintf(f, "%.10f\n", weights[c][i][j]);
            }
        }
    }
    fclose(f);
    
    f = fopen("bias_layer8_new.txt", "w");
    if (f) {
        fprintf(f, "%.10f", bias);
        fclose(f);
    }
}

int main() {
    load_data();
    init_weights();
    
    printf("Training for %d iterations...\n", NUM_ITERATIONS);
    
    double loss_sum = 0;
    int report_interval = 100000;
    
    for (int iter = 0; iter < NUM_ITERATIONS; iter++) {
        // Pick random sample
        int f_idx = rand() % FRAMES;
        int u = rand() % TARGET_ROWS;
        int v = rand() % TARGET_COLS;
        
        // Target
        double target = targets[f_idx * TARGET_ROWS * TARGET_COLS + u * TARGET_COLS + v];
        
        // Forward Pass with REAL Race Condition (OpenMP)
        double pred = bias;
        
        // Precompute coordinate ranges
        int start_kr = (u % 2 == 0) ? 0 : 1;
        int start_kc = (v % 2 == 0) ? 0 : 1;
        
        #pragma omp parallel for shared(pred)
        for (int c = 0; c < CHANNELS; c++) {
            double sum_c = 0;
            for (int kr = start_kr; kr < KERNEL_SIZE; kr += 2) {
                int input_row = (u - kr) / 2;
                if (input_row < 0 || input_row >= ROWS) continue;
                
                for (int kc = start_kc; kc < KERNEL_SIZE; kc += 2) {
                    int input_col = (v - kc) / 2;
                    if (input_col < 0 || input_col >= COLS) continue;
                    
                    double val = inputs[(long)f_idx * CHANNELS * ROWS * COLS + 
                                        c * ROWS * COLS + 
                                        input_row * COLS + input_col];
                    sum_c += val * weights[c][kr][kc];
                }
            }
            // NO ATOMIC, NO CRITICAL - INTENTIONAL RACE CONDITION
            pred += sum_c; 
        }
        
        double error = pred - target; 
        loss_sum += error * error;
        
        // Update Weights and Bias
        bias -= LR * error;
        
        // The update is serial because we want stable updates based on the corrupted error
        for (int c = 0; c < CHANNELS; c++) {
            for (int kr = start_kr; kr < KERNEL_SIZE; kr += 2) {
                int input_row = (u - kr) / 2;
                if (input_row < 0 || input_row >= ROWS) continue;
                
                for (int kc = start_kc; kc < KERNEL_SIZE; kc += 2) {
                    int input_col = (v - kc) / 2;
                    if (input_col < 0 || input_col >= COLS) continue;
                    
                    double val = inputs[(long)f_idx * CHANNELS * ROWS * COLS + 
                                        c * ROWS * COLS + 
                                        input_row * COLS + input_col];
                    
                    weights[c][kr][kc] -= LR * error * val;
                }
            }
        }
        
        if ((iter + 1) % report_interval == 0) {
            printf("Iter %d, MSE: %.6f\n", iter + 1, loss_sum / report_interval);
            loss_sum = 0;
        }
    }
    
    save_weights();
    printf("Training done.\n");
    return 0;
}
